File: 06_sorting/number_of_disc_intersections.py
def solution(a):
  n = len(a)
  left_ends = []
  right_ends = []

  for i in xrange(n):
    left_ends.append(i - a[i])
    right_ends.append(i + a[i])

  left_ends.sort()
  right_ends.sort()

  cnt = -1
  ind = 0
  ret = 0

  for rpt in right_ends:
    while ind < n and left_ends[ind] <= rpt:
      cnt += 1
      ind += 1
    ret += cnt
    if ret > 10000000:
      return -1
    cnt -= 1

  return ret

File: 06_sorting/number_of_disc_intersections.cpp
#include "vector"
#include "algorithm"

using namespace std;

const int MAX = 1e7;

int solution(vector<int> &a) {
  int ret = 0;
  int n = a.size();
  vector<int> lpts(n, 0);
  vector<long long int> rpts(n, 0);
  int cnt = -1;
  int status = 0;
  for (int i = 0; i < n; i++) {
    lpts[i] = i - a[i];
    rpts[i] = (long long) i + a[i];
  }
  sort(lpts.begin(), lpts.end());
  sort(rpts.begin(), rpts.end());
  for (int i = 0; i < n; i++) {
    while (status < n && lpts[status] <= rpts[i]){
      cnt++;
      status++;
    }
    ret += cnt;
    if (ret > MAX) {
      return -1;
    }
    cnt--;
  }
  return ret;
}
